src/v/kafka/server/group_manager.cc

```diff
@@ -408,76 +408,37 @@ ss::future<> group_manager::recover_partition(
             group->reset_tx_state(term);
         }
     }
     p->term = term;
+    co_await ss::max_concurrent_for_each(
+      ctx.groups, group_batch_size, [this, term, p](auto& pair) {
+          return do_recover_group(
+            term, p, std::move(pair.first), std::move(pair.second));
+      });
+}
 
-    for (auto& [group_id, group_stm] : ctx.groups) {
-        if (group_stm.has_data()) {
-            auto group = get_group(group_id);
-            vlog(
-              klog.info,
-              "Recovering {} - {}",
-              group_id,
-              group_stm.get_metadata());
-            for (const auto& member : group_stm.get_metadata().members) {
-                vlog(
-                  klog.debug,
-                  "Recovering group {} member {}",
-                  group_id,
-                  member);
-            }
-            if (!group) {
-                group = ss::make_lw_shared<kafka::group>(
-                  group_id,
-                  group_stm.get_metadata(),
-                  _conf,
-                  p->partition,
-                  _tx_frontend,
-                  _feature_table,
-                  _serializer_factory(),
-                  _enable_group_metrics);
-                group->reset_tx_state(term);
-                _groups.emplace(group_id, group);
-                group->reschedule_all_member_heartbeats();
-            }
-
-            for (auto& [tp, meta] : group_stm.offsets()) {
-                group->try_upsert_offset(
-                  tp,
-                  group::offset_metadata{
-                    meta.log_offset,
-                    meta.metadata.offset,
-                    meta.metadata.metadata,
-                  });
-            }
-
-            for (auto& [id, epoch] : group_stm.fences()) {
-                group->try_set_fence(id, epoch);
-            }
-
-            for (auto& [id, txseq] : group_stm.tx_seqs()) {
-                group->try_set_tx_seq(id, txseq);
-            }
-
-            for (auto& [id, timeout] : group_stm.timeouts()) {
-                group->try_set_timeout(id, timeout);
-            }
+ss::future<> group_manager::do_recover_group(
+  model::term_id term,
+  ss::lw_shared_ptr<attached_partition> p,
+  group_id group_id,
+  group_stm group_stm) {
+    if (group_stm.has_data()) {
+        auto group = get_group(group_id);
+        vlog(
+          klog.info, "Recovering {} - {}", group_id, group_stm.get_metadata());
+        for (const auto& member : group_stm.get_metadata().members) {
+            vlog(klog.debug, "Recovering group {} member {}", group_id, member);
         }
-    }
 
-    for (auto& [group_id, group_stm] : ctx.groups) {
-        if (group_stm.prepared_txs().size() == 0) {
-            continue;
-        }
-        auto group = get_group(group_id);
         if (!group) {
             group = ss::make_lw_shared<kafka::group>(
               group_id,
-              group_state::empty,
+              group_stm.get_metadata(),
```

Comment:
if you position the moved code in the file further away from where it is moving then the diff is often far cleaner. for example here diff is intermingling things and it's hard to read.